adding preview attached image & upload from clipboard

floatinghotpot opened the pull request:
Added 2 features:

1. Previously, the attached image thumbnail is too small to view. Now clicking the thumbnail will display a popup to view the full-size image. If the image is larger than 300px, the popup will also be centered.

2. Previously, only support uploading from file system. Now also support uploading image from clipboard. Just press Ctrl-V to paste to preview, then click "clipboard", a image with name "clipboard.png" will be uploaded.

mquandalle commented on client/components/cards/attachments.jade:
```diff
@@ -3,6 +3,13 @@ template(name="cardAttachmentsPopup")
       input.js-attach-file.hide(type="file" name="file" multiple)
       a.js-computer-upload {{_ 'computer'}}
+    li
+      a.js-upload-clipboard-image {{_ 'clipboard'}}
```
You should add the corresponding key in the `i18n/en.i18n.json` file (sorted by alphabetical order).

mquandalle commented on client/components/cards/attachments.jade:
```diff
@@ -3,6 +3,13 @@ template(name="cardAttachmentsPopup")
+      img.preview-clipboard-image(src="")
+
+template(name="previewAttachedImagePopup")
+  p {{_ "preview"}}
```
Same here, add the i18n key.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -1,5 +1,12 @@
 Template.attachmentsGalery.events({
-  'click .js-add-attachment': Popup.open('cardAttachments'),
+  'click .js-add-attachment'() {
+    Popup.open('cardAttachments').apply(this, arguments);
```
This should use ES6:

```es6
  'click .js-add-attachment'(...args) {
    Popup.open('cardAttachments').call(this, ...args);
```

floatinghotpot replied:
Good advice, I will try and verify. 

BTW, some ES6 rules seems not working in my Chrome, like string template, weird.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -1,5 +1,12 @@
+    Popup.open('cardAttachments').apply(this, arguments);
+
+    $('html').pasteImageReader(function(results) {
+      Popup.pasted = results;
```
I’m a bit confused about this `Popup.pasted`, could you add a comment on it

floatinghotpot replied:
The paste can only be fired by keyboard Ctrl-V. Here we temporarily store the content in Popup.pasted, later we get from it then upload.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
 });
+
+// ------------------------------------------------------------------------
+// Created by STRd6
```
You should move this code in a dedicated file, also I would recommend to keep it unmodified. Adding

```js
/*eslint-disable */
```

at the beginning of the file should disabling Eslint for it.

floatinghotpot replied:
Shall I put the separated code under /lib? where to put the link to it please?

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
+
+    setTimeout(function(){
+      const img = $('img.preview-large-image');
+      const w = img.width(), h = img.height();
```
Please define a single variable per line (repeat `const`).

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
+          top: (window.innerHeight - h)/2,
+        });
+      }
+    }, 50);
```
Could you explain the reason for this delay — and is there any chance we could remove it at some point?

floatinghotpot replied:
When there are multiple thumbnail, if click one then another very fast, will get a wrong width value. Adding a little delay will avoid the bug.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -20,6 +27,28 @@ Template.attachmentsGalery.events({
+    setTimeout(function(){
+      const img = $('img.preview-large-image');
+      const w = img.width(), h = img.height();
+      if(w > 300) {
```
We put a space after `if`, I’ll see how to add that in the linting rules.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
+      const card = this;
+      const file = new FS.File( results.file );
+      if(typeof results.file.type === 'string') {
+        const extname = results.file.type.split('/')[1];
```
IIRC CollectionFS has a method to extract a file extension name.

floatinghotpot replied:
results.file.type is usually "image/png", not a filename.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
+      if(typeof results.file.type === 'string') {
+        const extname = results.file.type.split('/')[1];
+        const filename = 'clipboard.png'.replace('png', extname);
+        file.name( filename );
```
Style: no space inside the parents

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
+      Attachments.insert(file);
+    }
+
+    $('html').pasteImageReader(function(){});
```
Could you add a comment about what this does — I guess, reset?

floatinghotpot replied:
Yes, clean up the Ctrl-V handler, as only allow it working in preview popup.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +67,77 @@ Template.cardAttachmentsPopup.events({
+      }
+      file.updatedAt(new Date());
+      file.boardId = card.boardId;
+      file.cardId  = card._id;
```
Style: don’t align `=`

mquandalle commented:
Hi @floatinghotpot,

Thank you for this PR, it looks great and I’ll be happy to merge it. I’ve written some relatively easy to fix (or answer) inline comments, and I’ll appreciate if you could take a look at them.

I also think that the preview should be on a separate popup displayed after the click on “Clipboard” (`Popup.open("clipboardPreview")`) and have a small text that explain what it does — “The following image is available in you clipboard, do you whish to attach it to this card” (or any better wording really).

mquandalle commented:
And we should also handle the case when there is no image the clipboard correctly — currently it only displays a broken image.

floatinghotpot commented:
Hi @mquandalle, thanks for the precise comments. I like the style. Will improve accordingly.

floatinghotpot commented:
Screenshot:
---
Menu:
![menu](https://cloud.githubusercontent.com/assets/2339512/10987598/c898ab4a-846f-11e5-9db8-f060e4909d79.png)
---
Preview:
![preview](https://cloud.githubusercontent.com/assets/2339512/10987602/cca5afbc-846f-11e5-942c-d873c13a67ca.png)

floatinghotpot commented:
Hi @mquandalle, I've fixed according to your comments. 
And, also add 2 more: 
1. not only allow paste image from clipboard, but also drag & drop image file.
2. translated zh-CN.

mquandalle commented:
Thanks for improving your PR, I’ll do a second review pass. However I’m not managing non-English translations via GitHub PRs (that would be difficult) but instead we use Transifex. So I encourage you to upload your Chinese file to Transifex (they may be some updates there since the last time I pulled the translations, I do that before each release), and could you please remove them from these PR (you can `git --amend`, `git push -f`).

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -20,6 +20,33 @@ Template.attachmentsGalery.events({
+Template.previewAttachedImagePopup.onRendered(function() {
+  // when there are multiple thumbnails,
+  // if click one then another very fast, we will get a wrong width value.
+  // adding a little delay will avoid the bug.
```
Could you try removing the `setTimeout` and executing `Tracker.flush()` before measuring the image to see if it works in the edge case you are mentioning?

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
+
+Template.previewClipboardImagePopup.onRendered(function() {
+  // we can paste image from clipboard
+  $('html').pasteImageReader(function(results) {
```
I think `$(document.body)` is a more accepted convention, right?

mquandalle commented on client/lib/dropImage.js:
```diff
@@ -0,0 +1,63 @@
+    });
+  };
+})(jQuery);
+
```
Please remove the extra line

mquandalle commented on client/lib/pasteImage.js:
```diff
@@ -0,0 +1,58 @@
+    });
+  };
+})(jQuery);
+
```
same

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
+    const results = Popup.pasted;
+    if(results && results.file) {
+      const card = this;
+      const file = new FS.File( results.file );
```
Style: no space between parents

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
+    if(results && results.file) {
+      const card = this;
+      const file = new FS.File( results.file );
+      if(!results.name) {
```
Style: a space after `if`

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
+  // we can also drag & drop image file to it
+  $('html').dropImageReader(function(results) {
+    if(results.dataURL.startsWith('data:image/')) {
+      Popup.pasted = results;
```
You should use a local variable and not pollute the Popup object (which is too fragile and will cause bugs). Just create something `let droppedImage;` on top of this file, and use it here.

mquandalle commented on client/components/cards/attachments.js:
```diff
@@ -38,4 +65,47 @@ Template.cardAttachmentsPopup.events({
+      file.cardId = card._id;
+      Attachments.insert(file);
+
+      delete Popup.pasted;
```
to remove

floatinghotpot commented:
Hi @mquandalle, I've improved the PR again, hope this time can pass your review.

BTW, besides this PR, in my devel branch, I've implemented some features for our production purpose in real project:
* 7c47384 add: import members/labels/lists as template from other board
* 63b9a64 add: import card from other board list
* 869735e allow creating multi cards by pasting excel tsv with field 'title','desc','duedate','member','label'
* 297f5e9 add task date to card, auto record date when drop to list, if status defined for list
* 7f84285 extend card and list model for task date & status
* 4a3525b improve PR, allow paste / drag&drop to upload
* 955fabf add preview attached image & allow upload image from clipboard

I can create separated PR with each feature, are you interested?

mquandalle commented:
Thanks for improving your PR!

I’m very confused by the broken image by default:

![wekan-attachments](https://cloud.githubusercontent.com/assets/1730702/11022002/3add1dbe-8608-11e5-8bb9-368b5128df4e.png)

Could you remove it if there is no valid image to display. Also I would to instructions on first line (before the image) and add some vertical margin. The button should also have the class name `.primary`.

floatinghotpot commented:
I removed the default src="", which is white blank in my chrome.

Now changed to:
![screen](https://cloud.githubusercontent.com/assets/2339512/11036208/103de658-8732-11e5-81fa-8706fb6e2eda.png)

floatinghotpot commented:
@mquandalle I've merged the changes into 1 commit. Is there anything else for me to improve in this PR?

mquandalle commented:
It’s all good, I expect to merge your PR this week-end :)

floatinghotpot commented:
@mquandalle I am sorry to hear about the terrible things happened in Paris. Hope you are doing well.

mquandalle commented:
Yes, my family and friends are safe. I’m not currently in France (I’m in San Francisco) but I spent a lot of time watching the news and calling my relatives this week-end and I haven’t worked at all.
